#include "mainwindow.h"
#include "ui_mainwindow.h"

#include <algorithm>
#include <random>
#include <functional> // review value -> std::less (Düôè)

namespace {
static constexpr int kMaxDequeSize = 1000;

template <class ModelT>
inline void moveCursorToBeginOrEnd(ModelT& model) {
    if (model.items.empty()) {
        model.toEnd();
    } else {
        model.toBegin();
    }
}
}

const std::deque<QString> MainWindow::tea = {
    "–ß–∞–π –õ—É–Ω—Ü–∑–∏–Ω", "–≠—Ä–ª –ì—Ä–µ–π", "–°–µ–Ω—á–∞", "–ü—É—ç—Ä", "–î–∞—Ä–¥–∂–∏–ª–∏–Ω–≥",
    "–ê—Å—Å–∞–º", "–ú–∞—Ç—á–∞", "–ì–∞–Ω–ø–∞—É–¥–µ—Ä", "–û–æ–ª–æ–Ω–≥", "–õ–∞–ø—Å–∞–Ω–≥ –°—É—à–æ–Ω–≥"
};

const std::deque<QString> MainWindow::cakes = {
    "–ö—Ä–∞—Å–Ω—ã–π –±–∞—Ä—Ö–∞—Ç", "–ù–∞–ø–æ–ª–µ–æ–Ω", "–ú–µ–¥–æ–≤–∏–∫", "–¢–∏—Ä–∞–º–∏—Å—É", "–ü—Ä–∞–≥–∞",
    "–ß–∏–∑–∫–µ–π–∫", "–ó–∞—Ö–µ—Ä", "–≠—Å—Ç–µ—Ä—Ö–∞–∑–∏", "–ú–æ—Ä–∫–æ–≤–Ω—ã–π —Ç–æ—Ä—Ç", "–ß—ë—Ä–Ω—ã–π –ª–µ—Å"
};

MainWindow::MainWindow(QWidget* parent)
    : QMainWindow(parent),
    ui(new Ui::MainWindow) {
    ui->setupUi(this);
    applyModel();
}

MainWindow::~MainWindow() {
    delete ui;
}

// –¢–µ–∫—É—â–∏–π –∫–æ–º–ø–∞—Ä–∞—Ç–æ—Ä –ø–æ —Ä–µ–∂–∏–º—É —Å–æ—Ä—Ç–∏—Ä–æ–≤–∫–∏
MainWindow::Cmp MainWindow::currentComp() const {
    static auto caseInsensitiveLess = [](const QString& a, const QString& b) {
        return a.compare(b, Qt::CaseInsensitive) < 0;
    };

    switch (sortMode_) {
    case SortMode::CaseInsensitive:
        return caseInsensitiveLess;      // —Å–≤–æ–π –∫–æ–º–ø–∞—Ä–∞—Ç–æ—Ä (CI)
    case SortMode::CaseSensitive:
        return std::less<QString>{};     // —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–π (CSüòâ)
    case SortMode::None:
    default:
        return std::less<QString>{};     // –¥–µ—Ñ–æ–ª—Ç CSüòâ)
    }
}

void MainWindow::applyModel() {
    QSignalBlocker blocker(ui->list_widget);
    ui->list_widget->setUpdatesEnabled(false);
    ui->list_widget->clear();

    for (int i = 0; i < static_cast<int>(deque_model_.items.size()); ++i) {
        ui->list_widget->addItem(QString("%1: %2").arg(i).arg(deque_model_.items[i]));
    }
    ui->list_widget->addItem("end");

    ui->txt_size->setText(QString::number(deque_model_.items.size()));

    deque_model_.clamp();

    const int currentRowIndex = deque_model_.atEnd()
                                    ? static_cast<int>(deque_model_.items.size())
                                    : static_cast<int>(deque_model_.pos);

    ui->list_widget->setCurrentRow(currentRowIndex);

    ui->list_widget->setUpdatesEnabled(true);
    updateControls();
    applyIterator();
}

void MainWindow::applyIterator() {
    const bool atEnd = deque_model_.atEnd();

    ui->txt_elem_content->setText(atEnd ? "" : deque_model_.items[deque_model_.pos]);

    // prev –∞–∫—Ç–∏–≤–Ω–∞ –≤–µ–∑–¥–µ, –∫—Ä–æ–º–µ –Ω–∞—á–∞–ª–∞ (–∏ –∫–æ–≥–¥–∞ –ø—É—Å—Ç–æ)
    const bool isEmpty  = deque_model_.items.empty();
    const bool atBegin  = !atEnd && (deque_model_.pos == 0);
    ui->btn_prev->setEnabled(!isEmpty && !atBegin);

    // next –∞–∫—Ç–∏–≤–Ω–∞, –ø–æ–∫–∞ –Ω–µ –Ω–∞ end
    ui->btn_next->setEnabled(!atEnd);

    ui->btn_edit->setEnabled(!atEnd);
    ui->btn_erase->setEnabled(!atEnd);

    // —Å–∏–Ω—Ö—Ä–æ–Ω–∏–∑–∞—Ü–∏—è –≤—ã–¥–µ–ª–µ–Ω–∏—è
    const int currentRowIndex = atEnd
                                    ? static_cast<int>(deque_model_.items.size())
                                    : static_cast<int>(deque_model_.pos);

    if (ui->list_widget->currentRow() != currentRowIndex) {
        ui->list_widget->setCurrentRow(currentRowIndex);
    }
}

void MainWindow::updateControls() {
    const bool isEmpty = deque_model_.items.empty();

    ui->btn_pop_front->setEnabled(!isEmpty);
    ui->btn_pop_back->setEnabled(!isEmpty);
    ui->btn_erase->setEnabled(!isEmpty && !deque_model_.atEnd());

    // –º–∏–º–∏–∫—Ä–∏—è
    ui->btn_lower_bound->setEnabled(true);
    ui->btn_upper_bound->setEnabled(true);
    ui->btn_unique->setEnabled(true);

}

// === –ë–∞–∑–æ–≤—ã–µ –æ–ø–µ—Ä–∞—Ü–∏–∏ ===

void MainWindow::on_btn_push_front_clicked() {
    const QString text = ui->txt_elem_content->text().trimmed();
    if (text.isEmpty()) {
        return;
    }

    deque_model_.items.push_front(text);
    sortMode_ = SortMode::None;
    moveCursorToBeginOrEnd(deque_model_);
    applyModel();
}

void MainWindow::on_btn_push_back_clicked() {
    const QString text = ui->txt_elem_content->text().trimmed();
    if (text.isEmpty()) {
        return;
    }

    deque_model_.items.push_back(text);
    sortMode_ = SortMode::None;
    moveCursorToBeginOrEnd(deque_model_);
    applyModel();
}

void MainWindow::on_btn_pop_front_clicked() {
    if (deque_model_.items.empty()) {
        return;
    }

    deque_model_.items.pop_front();
    sortMode_ = SortMode::None;
    moveCursorToBeginOrEnd(deque_model_);
    applyModel();
}

void MainWindow::on_btn_pop_back_clicked() {
    if (deque_model_.items.empty()) {
        return;
    }

    deque_model_.items.pop_back();
    sortMode_ = SortMode::None;
    moveCursorToBeginOrEnd(deque_model_);
    applyModel();
}

void MainWindow::on_btn_clear_clicked() {
    deque_model_.items.clear();
    sortMode_ = SortMode::None;
    deque_model_.toEnd();
    applyModel();
}

void MainWindow::on_btn_insert_clicked() {
    const QString text = ui->txt_elem_content->text().trimmed();
    if (text.isEmpty()) {
        return;
    }

    const size_t insertIndex = deque_model_.atEnd()
                                   ? deque_model_.items.size()
                                   : deque_model_.pos;

    auto it = deque_model_.items.begin();
    std::advance(it, static_cast<long>(insertIndex));
    deque_model_.items.insert(it, text);

    sortMode_ = SortMode::None;
    moveCursorToBeginOrEnd(deque_model_);
    applyModel();
}

void MainWindow::on_btn_erase_clicked() {
    if (deque_model_.atEnd()) {
        return;
    }

    auto it = deque_model_.items.begin();
    std::advance(it, static_cast<long>(deque_model_.pos));
    deque_model_.items.erase(it);

    sortMode_ = SortMode::None;
    moveCursorToBeginOrEnd(deque_model_);
    applyModel();
}

void MainWindow::on_btn_edit_clicked() {
    if (deque_model_.atEnd()) {
        return;
    }

    const QString text = ui->txt_elem_content->text().trimmed();
    if (text.isEmpty()) {
        return;
    }

    // –ø—Ä–∞–≤–∏–º —Ç–µ–∫—É—â–µ–µ –∑–Ω–∞—á–µ–Ω–∏–µ, –ø–æ–∑–∏—Ü–∏—é –ù–ï —Ç—Ä–æ–≥–∞–µ–º
    deque_model_.items[deque_model_.pos] = text;
    sortMode_ = SortMode::None;
    applyModel();
}

// === –ù–∞–≤–∏–≥–∞—Ü–∏—è ===

void MainWindow::on_btn_prev_clicked() {
    if (deque_model_.items.empty()) {
        return;
    }

    if (!deque_model_.atEnd() && deque_model_.pos > 0) {
        --deque_model_.pos;
    } else if (deque_model_.atEnd()) {
        // —Å end —à–∞–≥–∞–µ–º –Ω–∞ –ø–æ—Å–ª–µ–¥–Ω–∏–π —ç–ª–µ–º–µ–Ω—Ç
        deque_model_.pos = deque_model_.items.size() - 1;
    }

    applyIterator();
}

void MainWindow::on_btn_next_clicked() {
    if (!deque_model_.atEnd()) {
        ++deque_model_.pos;
        applyIterator();
    }
}

void MainWindow::on_btn_begin_clicked() {
    deque_model_.toBegin();
    applyIterator();
}

void MainWindow::on_btn_end_clicked() {
    if (deque_model_.items.empty()) {
        deque_model_.toEnd();
    } else {
        // —à–∞–≥–Ω—É—Ç—å –∫ —Ñ–∏–∫—Ç–∏–≤–Ω–æ–º—É end
        deque_model_.pos = deque_model_.items.size();
    }
    applyIterator();
}

void MainWindow::on_list_widget_currentRowChanged(int currentRow) {
    currentRow = std::clamp(currentRow, 0, static_cast<int>(deque_model_.items.size()));
    deque_model_.pos = static_cast<size_t>(currentRow);
    applyIterator();
}

// === –ò–∑–º–µ–Ω–µ–Ω–∏–µ —Ä–∞–∑–º–µ—Ä–∞ –∏ –ø–µ—Ä–µ—Å—Ç–∞–Ω–æ–≤–∫–∏ ===

void MainWindow::on_btn_resize_clicked() {
    bool ok = false;
    const int newSizeInt = ui->txt_size->text().toInt(&ok);
    if (!ok || newSizeInt < 0 || newSizeInt > kMaxDequeSize) {
        return;
    }

    const size_t newSize = static_cast<size_t>(newSizeInt);
    deque_model_.items.resize(newSize);
    sortMode_ = SortMode::None;

    // –ø–æ—Å–ª–µ resize –∫—É—Ä—Å–æ—Ä –Ω–∞ –Ω–∞—á–∞–ª–æ (–∏–ª–∏ end, –µ—Å–ª–∏ –ø—É—Å—Ç–æ)
    if (newSize == 0) {
        deque_model_.toEnd();
    } else {
        deque_model_.toBegin();
    }

    applyModel();
}

void MainWindow::on_btn_reverse_clicked() {
    if (deque_model_.items.empty()) {
        return;
    }

    // —Å–æ—Ö—Ä–∞–Ω—è–µ–º –∏–Ω–¥–µ–∫—Å
    const size_t savedIndex = deque_model_.atEnd()
                                  ? deque_model_.items.size()
                                  : deque_model_.pos;

    std::reverse(deque_model_.items.begin(), deque_model_.items.end());
    sortMode_ = SortMode::None;

    deque_model_.pos = std::min(savedIndex, deque_model_.items.size());
    applyModel();
}

void MainWindow::on_btn_shuffle_clicked() {
    if (deque_model_.items.empty()) {
        return;
    }

    // —Å–æ—Ö—Ä–∞–Ω—è–µ–º –∏–Ω–¥–µ–∫—Å
    const size_t savedIndex = deque_model_.atEnd()
                                  ? deque_model_.items.size()
                                  : deque_model_.pos;

    std::shuffle(deque_model_.items.begin(), deque_model_.items.end(), random_gen_);
    sortMode_ = SortMode::None;

    deque_model_.pos = std::min(savedIndex, deque_model_.items.size());
    applyModel();
}

// === –ü–æ–∏—Å–∫/–ø–æ–¥—Å—á—ë—Ç/–º–∏–Ω-–º–∞–∫—Å ===

void MainWindow::on_btn_find_clicked() {
    const QString key = ui->txt_elem_content->text().trimmed();
    if (key.isEmpty()) {
        return;
    }

    auto it = std::find(deque_model_.items.begin(), deque_model_.items.end(), key);
    if (it == deque_model_.items.end()) {
        // –Ω–µ –Ω–∞–π–¥–µ–Ω–æ ‚Äî –∫—É—Ä—Å–æ—Ä –Ω–∞ end
        deque_model_.toEnd();
    } else {
        deque_model_.pos = static_cast<size_t>(std::distance(deque_model_.items.begin(), it));
    }

    applyModel();
}

void MainWindow::on_btn_count_clicked() {
    const QString key = ui->le_count->text().trimmed();
    if (key.isEmpty()) {
        ui->lbl_count->setText("0");
        return;
    }

    const int countValue = static_cast<int>(
        std::count(deque_model_.items.begin(), deque_model_.items.end(), key)
        );

    ui->lbl_count->setText(QString::number(countValue));
    // –∫—É—Ä—Å–æ—Ä –Ω–µ —Ç—Ä–æ–≥–∞–µ–º
}

void MainWindow::on_btn_min_element_clicked() {
    if (deque_model_.items.empty()) return;
    auto less = currentComp();
    auto it = std::min_element(deque_model_.items.begin(), deque_model_.items.end(), less); // —Å—Ç–∞–Ω–¥ (CSüòâ)
    deque_model_.pos = static_cast<size_t>(std::distance(deque_model_.items.begin(), it));
    applyModel();
}

void MainWindow::on_btn_max_element_clicked() {
    if (deque_model_.items.empty()) return;
    auto less = currentComp();
    auto it = std::max_element(deque_model_.items.begin(), deque_model_.items.end(), less); // —Å—Ç–∞–Ω–¥ (CSüòâ)
    deque_model_.pos = static_cast<size_t>(std::distance(deque_model_.items.begin(), it));
    applyModel();
}


// === –°–æ—Ä—Ç–∏—Ä–æ–≤–∫–∞ –∏ –ø—Ä–æ–∏–∑–≤–æ–¥–Ω—ã–µ –æ–ø–µ—Ä–∞—Ü–∏–∏ ===

void MainWindow::on_btn_merge_sort_clicked() {
    sortMode_ = SortMode::CaseSensitive;
    deque_model_.items = MergeSort(deque_model_.items, currentComp());
    deque_model_.toBegin(); // –Ω–∞ —Å–æ—Ä—Ç–∏—Ä–æ–≤–∫—É –∫—É—Ä—Å–æ—Ä –≤ 0
    applyModel();
}

void MainWindow::on_btn_merge_sOrT_clicked() {
    sortMode_ = SortMode::CaseInsensitive;
    deque_model_.items = MergeSort(deque_model_.items, currentComp());
    deque_model_.toBegin(); // –Ω–∞ —Å–æ—Ä—Ç–∏—Ä–æ–≤–∫—É –∫—É—Ä—Å–æ—Ä –≤ 0
    applyModel();
}

void MainWindow::on_btn_unique_clicked() {
    if (deque_model_.items.empty() || sortMode_ == SortMode::None) {
        return;
    }

    auto less = currentComp();
    auto equalByLess = [less](const QString& a, const QString& b) {
        return !less(a, b) && !less(b, a);
    };

    auto newEnd = std::unique(deque_model_.items.begin(), deque_model_.items.end(), equalByLess);
    deque_model_.items.erase(newEnd, deque_model_.items.end());
    deque_model_.toBegin(); // unique –Ω–∞—á–∞–ª–æ
    applyModel();
}

void MainWindow::on_btn_lower_bound_clicked() {
    const QString key = ui->txt_elem_content->text().trimmed();
    if (key.isEmpty() || sortMode_ == SortMode::None) {
        return;
    }

    auto it = std::lower_bound(deque_model_.items.begin(), deque_model_.items.end(), key, currentComp());
    deque_model_.pos = static_cast<size_t>(std::distance(deque_model_.items.begin(), it)); // –º–æ–∂–µ—Ç –±—ã—Ç—å end
    applyModel();
}

void MainWindow::on_btn_upper_bound_clicked() {
    const QString key = ui->txt_elem_content->text().trimmed();
    if (key.isEmpty() || sortMode_ == SortMode::None) {
        return;
    }

    auto it = std::upper_bound(deque_model_.items.begin(), deque_model_.items.end(), key, currentComp());
    deque_model_.pos = static_cast<size_t>(std::distance(deque_model_.items.begin(), it)); // –º–æ–∂–µ—Ç –±—ã—Ç—å end
    applyModel();
}

// === –ù–∞–±–æ—Ä—ã ===

void MainWindow::on_btn_tea_clicked() {
    deque_model_.items = tea;
    deque_model_.toBegin();
    sortMode_ = SortMode::None;
    applyModel();
}

void MainWindow::on_btn_cakes_clicked() {
    deque_model_.items = cakes;
    deque_model_.toBegin();
    sortMode_ = SortMode::None;
    applyModel();
}
